MaxText/layers/pipeline.py

```diff
@@ -0,0 +1,322 @@
+    mesh:  The device mesh of the system.
+    remat_policy: Remat policy to use for the loop iterations
+  """
+  
+  config: common_types.Config
+  layers: nn.Module # The name of this property (layers) is reflected in the state pytree and thus also checkpoints.
+  mesh: common_types.Mesh
+  remat_policy: Any = None
+
+  def setup(self):
+    self.num_stages = self.config.ici_pipeline_parallelism * self.config.dcn_pipeline_parallelism
+    self.microbatch_size = self.config.global_batch_size_to_train_on // self.config.num_pipeline_microbatches
+    microbatches_per_stage = self.config.num_pipeline_microbatches // self.num_stages
+    self.microbatches_per_stage = microbatches_per_stage
+
+  def init_states(self, inputs):
+    '''Initialize components of state: state_io, shift
+        Assumes input has already been reshaped into microbatches: [num_micro_batches, micro_batch_size, sequence, embed]
+
+        Returns a dictionary with properties
+          shift: zeros shape [num_stages, micro_size, sequence, embed]
+          state_io: reshaped inputs [num_stages, microbatches/stages, micro_size, sequence, embed]
+          loop_iteration: scalar set initially to 0.  
+    '''
+
+    # Shift is used to rotate the output of each pipeline into the input of the next
+    # shift has shape [num_stages, micro_size, sequence, embed]
+    shift = jnp.zeros((self.num_stages,) + inputs.shape[1:], dtype=inputs.dtype)
+    shift = nn.with_logical_constraint(shift, ("activation_stage", "activation_batch", "activation_length", "activation_embed"),rules=self.config.logical_axis_rules,mesh=self.mesh)
+
+    # state_io (state input output) at first holds all of the input batches, but also will hold the outputs as the pipeline runs/finishes
+    # state_io has shape [num_stages, microbatches/stages, micro_size, sequence, embed]
+    state_io = jnp.reshape(inputs, (self.num_stages, self.microbatches_per_stage) + inputs.shape[1:])
+    # We shard the microbatch_size axis by data/fsdp, not num_microbatches since those are looped over.
+    state_io = nn.with_logical_constraint(state_io, ("activation_stage", None, "activation_batch", "activation_length", "activation_embed"),rules=self.config.logical_axis_rules, mesh=self.mesh) 
+
+    init_loop_state = {
+      "state_io": state_io,
+      "shift": shift,
+      "loop_iteration": 0
+    }
+    return init_loop_state
+
+  def get_iteration_inputs(self, loop_iteration, state_io, shift):
+    '''
+    Construct stages_in: the global array that is operated on for this iteration, shape same as shift=[stages, micro_size, sequence, embed]
+    This is almost a rotated version of the last outputs, except for the first stage which must grab a new batch from state_io
+    '''
+
+    # Setup potential input from state_io, which has a rotating microbatch index (size of microbatches_per_stage)
+    state_io_batch_idx = loop_iteration % self.microbatches_per_stage
+    first_stage_in = state_io[:,state_io_batch_idx]
```

Comment:
@gobbleturk How do we prevent stage 0 from processing the output data during the bubble?

Reply:
It occurred to me that any output data processed will be discarded while still in the `shift` buffer, so it's a non-issue. Let me know if that's a correct understanding.

Reply:
Yes that is the correct understanding! Added a comment here similar to your point